src/java/org/apache/cassandra/service/CleanupStateTracker.java

```diff
@@ -0,0 +1,171 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class CleanupStateTracker
+{
+    private static final Logger log = LoggerFactory.getLogger(CleanupStateTracker.class);
+
+    @VisibleForTesting
+    static final String CLEANUP_STATE_FILE_NAME = "node_cleanup_state.json";
+    private static final Instant MIN_TS = Instant.parse("2022-01-01T00:00:00Z");
+
+    private final CleanupState cleanupStateCache;
+
+    public CleanupStateTracker() throws IOException
+    {
+        this.cleanupStateCache = new CleanupState(createOrRetrieveCleanupStateFile());
+    }
+
+    @VisibleForTesting
+    File createOrRetrieveCleanupStateFile() throws IOException
+    {
+        File cleanupStateFile = new File(cleanupStateFileLocation());
+        try
+        {
+            cleanupStateFile.createNewFile();
+            return cleanupStateFile;
+        }
+        catch (IOException e)
+        {
+            log.warn("Cannot retrieve or create node cleanup state file.", cleanupStateFile.getAbsolutePath(), e);
+            throw e;
+        }
+    }
+
+    @VisibleForTesting
+    String cleanupStateFileLocation(){
+        return DatabaseDescriptor.getPersistentSettingsLocation() + CLEANUP_STATE_FILE_NAME;
+    }
+
+    /** Creates table entry if it does not already exist */
+    public synchronized void createCleanupEntryForTableIfNotExists(String keyspaceName, String columnFamily)
+        throws IOException
+    {
+        if (!cleanupStateCache.entryExists(keyspaceName, columnFamily))
+            cleanupStateCache.updateTsForEntry(keyspaceName, columnFamily, MIN_TS);
+    }
+
+    /** Updates ts for table entry */
+    public synchronized void recordSuccessfulCleanupForTable(String keyspaceName, String columnFamily)
+        throws IOException
+    {
+        cleanupStateCache.updateTsForEntry(keyspaceName, columnFamily,Instant.now());
+    }
+
+    /** Returns min ts for this node, null if no entry exists. */
+    public Instant getLastSuccessfulCleanupTsForNode()
+    {
+        return cleanupStateCache.getMinimumTsOfAllEntries();
+    }
+
+    @VisibleForTesting
+    static class CleanupState
+    {
+        private final ConcurrentMap<String, Long> tableEntries;
+        private final File persistentCleanupStateFile;
+        private final ObjectMapper objectMapper = new ObjectMapper();
+
+        public CleanupState(File cleanupStateFile) throws IOException
```

Comment:
My recommendation is that you do any file-based operations (file creation, mapping, persistence) in one class.  Right now you have some operations in this inner class, and others in the class above.

A more logical separation would be for `CleanupState` to be only responsible for holding the in-memory representation of the data (i.e., the tableEntries), and then maybe a `CleanupStatePersister` which is responsible for writing or reading from disk.

Reply:
That makes sense, I'll make those changes